src/azure_hl.rs

```diff
@@ -569,9 +569,24 @@ impl DrawTarget {
             None => ptr::null_mut(),
             Some(ref mut draw_options) => draw_options as *mut AzDrawOptions
         };
+
+        let draw_rect = Rect(rect.origin,
+            match pattern {
+                PatternRef::Surface(surface) => {
+                    let surface_size = surface.size();
+                    match (surface.repeat_x, surface.repeat_y) {
+                        (true, true) => rect.size,
+                        (true, false) => Size2D(rect.size.width, surface_size.height as f32),
+                        (false, true) => Size2D(surface_size.width as f32, rect.size.height),
+                        (false, false) => Size2D(surface_size.width as f32, surface_size.height as f32),
+                    }
+                },
+                _ => rect.size,
+            }
+        );
         unsafe {
             AzDrawTargetFillRect(self.azure_draw_target,
-                                 &mut rect.as_azure_rect(),
+                                 &mut draw_rect.as_azure_rect(),
```

Comment:
Why not just pass `rect` in directly? I'd rather have the callers of `fill_rect` do the conversion from `rect` to `draw_rect` as necessary and leave this function more or less as is.